services/gateway/gateway-service.mjs

```diff
@@ -245,53 +250,84 @@ class GatewayService {
    * - displayName: The name to show for this player.
    * - displayChar: The character asset to display.
    */
-  async onUserLocation(socket, msg) {
+  async onUserLocation(socket, origMsg) {
+    const msg = {};
     msg.playerID = socket.playerID;
     msg.displayName = socket.playerData.displayName;
     msg.displayChar = socket.playerData.displayChar;
+    msg.facing = origMsg.facing;
+    msg.mapCoord = MapCoord.fromObject(origMsg.mapCoord);
     // TODO: Check if facing is valid.
     // TODO: Check if movement is legal.
 
-    let mapSize = this.gameMap.getMapSize();
-    let lastMsg = { x:socket.playerData.x , y:socket.playerData.y}
-    if(lastMsg.x === undefined || lastMsg.y === undefined){
-      lastMsg.x = 0;
-      lastMsg.y = 0;
+    let lastCoord = socket.playerData.mapCoord;
+    let mapSize = this.gameMap.getMapSize(lastCoord.mapName);
+    if(lastCoord.x === undefined || lastCoord.y === undefined || lastCoord.mapName === undefined){
+      // Shouldn't happen, log an error.
+      console.error(`Invalid lastCoord in onUserLocation ${lastCoord}`);
+      return;
     }
     if(!this._movementRequestSpeedCheck(socket.playerData)){
+      console.warn(`Player ${msg.playerID} is speeding`);
+      return;
+    }
+    if (!(lastCoord.mapName === msg.mapCoord.mapName)) {
+      console.warn(`Player ${msg.playerID} changed map from ` +
+        `${lastCoord.mapName} to ${msg.mapCoord.mapName} without permission.`);
```

Comment:
I think this check should be included in the speed check.

Reply:
Will leave this to whyang when he moves everything that can be moved into the library.